avm/src/vm/opcode.rs

```diff
@@ -3072,29 +3065,16 @@ pub static OPCODES: Lazy<HashMap<i64, ByteOpcode>> = Lazy::new(|| {
       eprintln!("Event transmission error");
       std::process::exit(3);
     }
-    // Get the HTTP responses lock and periodically poll it for responses from the user code
-    // TODO: Swap from a timing-based poll to getting the waker to the user code so it can be
-    // woken only once and reduce pressure on this lock.
+    // Get the HTTP responses lock and async poll it for responses from the user code
     let responses = Arc::clone(&HTTP_RESPONSES);
-    let response_hm = poll_fn(|cx: &mut Context<'_>| -> Poll<Arc<HandlerMemory>> {
-      let responses_hm = responses.lock().unwrap();
-      let hm = responses_hm.get(&conn_id);
-      if let Some(hm) = hm {
-        Poll::Ready(hm.clone())
+    loop {
+      if responses.get(&conn_id).is_some() {
+        break;
       } else {
-        drop(hm);
-        drop(responses_hm);
-        let waker = cx.waker().clone();
-        // TODO: os threads are quite expensive and could cause our runtime to
-        // stop executing, this might be a point of optimization:
-        thread::spawn(|| {
-          thread::sleep(Duration::from_millis(10));
-          waker.wake();
-        });
-        Poll::Pending
+        sleep(Duration::from_millis(10)).await;
       }
-    })
-    .await;
+    }
+    let response_hm = responses.get(&conn_id).unwrap();
```

Comment:
we should probably clear this field in `HTTP_RESPONSES` or a long running server could hit an OOM after a very long time?

Reply:
confirmed this is a memory leak and created https://github.com/alantech/meta/issues/171 to track it

Reply:
It's not a memory leak, though? If you have a huge buffer of requests, it will take memory regardless of the transit mechanism.

Reply:
What? We trigger a delete on send, don't we?